js/style/style.js

```diff
@@ -362,7 +369,16 @@ class Style extends Evented {
 
         this._layers[id] = layer;
 
-        delete this._removedLayers[id];
+        if (this._removedLayers[id]) {
+            // If, in the current batch, we have already removed this layer
+            // and we are now re-adding it, then we need to clear (rather
+            // than just reload) the underyling source's tiles.
+            // Otherwise, tiles marked 'reloading' will have buffers that are
+            // set up for the _previous_ version of this layer, confusing
+            // https://github.com/mapbox/mapbox-gl-js/issues/3633
+            delete this._removedLayers[id];
+            this._updatedSources[layer.source] = 'clear';
```

Comment:
Will these tiles still be marked as `reloading` within `SourceCache`? [As it is defined now, `reloading` means that there is an asynchronous update underway and the tiles are still renderable](https://github.com/mapbox/mapbox-gl-js/blob/master/js/source/source_cache.js#L139-L142).

Is there a need for another state that means there is an asynchronous update underway and the tiles are NOT renderable? Or is there a need to redefine `reloading`?